lib/protocol/websocket/close_frame.rb

```diff
@@ -29,11 +29,41 @@ class CloseFrame < Frame
 			def unpack
 				data = super
 				
-				return data.unpack(FORMAT)
+				case data.length
+				when 0
+					[nil, ""]
+				when 1
+					raise ProtocolError, "invalid close frame length!"
+				else
+					code, reason = *data.unpack(FORMAT)
+					
+					case code
+					when 0 .. 999, 1005 .. 1006, 1015, 5000 .. 0xFFFF
+						raise ProtocolError, "invalid close code!"
+					when 1004, 1016 .. 2999
+						raise ProtocolError, "reserved close code!"
+					end
+					
+					reason.force_encoding(Encoding::UTF_8)
+					
+					unless reason.valid_encoding?
+						raise ProtocolError, "invalid UTF-8 in close reason!"
+					end
+					
+					[code, reason]
+				end
 			end
 			
 			def pack(code, reason)
-				super [code, reason].pack(FORMAT)
+				if code
+					unless reason.encoding == Encoding::UTF_8
+						reason = reason.encode(Encoding::UTF_8)
```

Comment:
I don't think we need a conditional here since `encode(Encoding::UTF_8)` should be a no-op if it's already encoded correctly.

Reply:
Depends on what the meaning of no-op is, the conditional avoids a useless copy of the string.

```ruby
puts "abc".tap { |e| puts e.object_id }.encode("UTF-8").object_id
300
320
```